NeoML/src/Dnn/Dnn.cpp

```diff
@@ -821,6 +846,7 @@ static const int DnnVersion = 2000;
 
 void CDnn::Serialize( CArchive& archive )
 {
+	NeoAssertMsg( referenceDnnInfo == nullptr, "For ReferenceDnn serializing is restricted" );
```

Comment:
The more I look at these asserts (which can be found in almost every function), the more I think that "reference dnn" is not a CDnn. It should be a separate class without all the forbidden methods. For convenience it might have a common ancestor with CDnn.

Reply:
I'm totally agree. It may be one of parent classes of `CDnn`, but one should copy almost all interface methods.

Reply:
But it is a significant interface change, should discuss with somebody else

Reply:
fixed

Reply:
> Please, don't do that. There is no need in inheritance here, it will certainly be an unnecessary complication. A lot of client code will require significant refactoring.
> 
> Current implementation looks quite reasonable as is.

No refactoring will be required unless the code uses "reference dnns", which is a new mechanism (no such production code exists). But without separating classes, ALL existing code will require additional checks whether a shoved "dnn" is a normal CDnn or defective reference

Reply:
And now there is no method to check whether CDnn is normal or reference

Reply:
As a proposal see https://github.com/neoml-lib/neoml/pull/1087

Reply:
Honestly, didn’t see convincing arguments for.

There won’t by any other inheritor classes except CDnn itself. Then why is this interface needed?  To my mind, this is an antipattern “inheritance abuse”, which results in nothing useful but unnecessary code complication.

Besides, NeoML is delivered as dynamic library with static linking. We don’t need interface-implementation separation of the code.

Without class separation we need to carefully incorporate creation of `Factory` and the calls to `CreateReferenceDnn`. Then the user of the mechanism/network doesn’t care whether he works with the reference or not. If it is not so, this is something wrong in design. With class separation, additionally, all type declarations of CDnn must be replaced to new interface — a lot of redundant changes and why? No significant reason.

Without class separation, all existing code needs checks — why is that? Only those networks, where parallelization is required. It is a task for the client who is responsible for parallelization.

Reply:
> the user of the mechanism/network doesn’t care whether he works with the reference or not

They **do** care. References cannot be even serialized (besides many other modifying methods). Separation provides static type-checking instead of asserts. 


> all existing code needs checks — why is that

In any _(at least public)_ function taking CDnn I would put `assert( !dnn.IsReference() )`

Reply:
No care. Main scenario for production code is as follows: the user deserializes ready-to-work network and applies it to the data given from outside. Modifying methods are not called on inference or just before it. 

You need modifying methods during building of network before first serialization. Usually it is done in separate utility application, where no parallelization is employed and certainly you can add assertions there (although this defend is also feels kinda useless).

Reply:
Current design with CReferenceDnn ok for me